soaktest/soakclient_transaction/soakclient_transaction.go

```diff
@@ -0,0 +1,552 @@
+					p,
+					c,
+					transactionCommitNum)
+				if err != nil {
+					fmt.Printf("Failed to commit transaction %v\n", err)
+					terminate(c, p, err, errorChan)
+					return
+				}
+				committed = true
+				p.BeginTransaction()
+			}
+		case ev := <-p.Events():
+			// Producer delivery report
+			switch e := ev.(type) {
+			case *kafka.Message:
+				soaktest.ProducerDeliveryCheck(e)
+
+			case *kafka.Stats:
+				err = soaktest.HandleStatsEvent(e, outTopic)
+				if err != nil {
+					fmt.Printf("Failed to handle stats event for "+
+						"producer %v\n", err)
+					terminate(c, p, err, errorChan)
+					return
+				}
+
+			case kafka.Error:
+				// kafka.Errors should generally be
+				// considered informational, the client
+				// will try to automatically recover.
+				soaktest.ProducerErr += 1
+				soaktest.DatadogIncrement(soaktest.PRODUCER_ERROR+
+					outTopic, 1)
+				fmt.Printf("kafka.Error: %v: %v for producer %v\n",
+					e.Code(), e, p)
+
+			default:
+				fmt.Printf("Ignored event: %s for producer %v\n", ev, p)
+			}
+
+		default:
+			var ev = c.Poll(100)
+			if ev == nil {
+				continue
+			}
+			switch e := ev.(type) {
+			case *kafka.Message:
+				if !soaktest.HandleMessage(e, hwmarks) {
+					fmt.Printf("Message receive failed, " +
+						"skip producing message")
+					continue
+				}
+				err = produceMessage(e, p, msgid, outTopic)
+				if err != nil {
+					fmt.Printf("Failed to Produce message to outTopic %s, %s\n",
+						outTopic, err)
+					continue
+				}
+				msgid += 1
+				position, err := getConsumerPosition(
+					c,
+					e.TopicPartition.Partition,
+					inputTopic)
+				if err != nil {
+					fmt.Printf("Failed to get consumer position: %v", err)
+					terminate(c, p, err, errorChan)
+					return
+				}
+				partitionPositionMap[e.TopicPartition.Partition] = position
+				if msgid%100 == 0 && !committed {
+					transactionCommitNum, err = commitTransaction(
+						getAllPartitionPositions(partitionPositionMap),
+						p,
+						c,
+						transactionCommitNum)
+					if err != nil {
+						fmt.Printf("Failed to commit transaction %s\n", err)
+						terminate(c, p, err, errorChan)
+						return
+					}
+					committed = true
+					p.BeginTransaction()
+				} else {
+					committed = false
+				}
+
+			case *kafka.Stats:
+				err = soaktest.HandleStatsEvent(e, inputTopic)
+				if err != nil {
+					fmt.Printf("Failed to handle stats event for "+
+						"consumer %v\n", err)
+					terminate(c, p, err, errorChan)
+					return
+				}
+
+			case kafka.Error:
+				// kafka.Errors should generally be
+				// considered informational, the client
+				// will try to automatically recover.
+				soaktest.ConsumerErr += 1
+				soaktest.DatadogIncrement(soaktest.CONSUMER_ERROR+
+					inputTopic, 1)
+				fmt.Printf("kafka.Error: %v: %v for consumer %v\n",
+					e.Code(), e, c)
+
+			default:
+				fmt.Printf("Ignored %v for consumer %v\n", e, c)
+			}
+		}
+	}
+	p.Close()
+	c.Close()
+}
```

Comment:
no wg signalling done here?

Reply:
Hi Magnus, thanks for pointing out this. Because we already added the wg signalling done from the caller side verifyProducerTransaction(), do we still need to add it here?